ssz/merkle-proofs.md

```diff
@@ -355,3 +359,113 @@ def verify_merkle_multiproof(leaves: Sequence[Bytes32],
 ```
 
 Note that the single-item proof is a special case of a multi-item proof; a valid single-item proof verifies correctly when put into the multi-item verification function (making the natural trivial changes to input arguments, `index -> [index]` and `leaf -> [leaf]`). Note also that `calculate_merkle_root` and `calculate_multi_merkle_root` can be used independently to compute the new Merkle root of a proof with leaves updated.
+
+### Dynamic multiproofs
+
+While generalized indices are useful for identifying specific elements of a merkle tree, they are inefficient when used directly to dynamically request specific leaves for a multiproof. This is because each generalized index must be encoded as a varint or some fixed-bitlength uint. N generalized indices would require somewhere between 2N to 8N bytes to be serialized.
+
+A more efficient multiproof descriptor encodes the "shape" a multiproof, rather than encoding the generalized indices individually. This shape can be encoded as a single bitlist that describes traversal through the merkle tree. Of note, in this format, leaf indices and helper indices are not treated separately, rather, the shape includes both and does not distinguish between them.
+
+We define the descriptor as a traversal through the merkle tree: starting from the root of the tree, traverse depth-first, left to right, encoding a zero bit if the node has a requested index as a descendent or a one bit if the node is a requested index.
+
+This equates to 2N - 1 bits total required to encode N leaf/witness locations: N one bits and N-1 zero bits.
+
+When the descriptor is serialized, up to 7 zero bits should be appended for byte-alignment.
+
+The proof leaves and witnesses should be ordered identically: depth-first, in-order.
```

Comment:
Is the "witness" same as the "helper" node? And the "requested" same as the "leaf"?
Usually witness means everything included in the proof, in which case "leaves and witnesses" makes no sense. Maybe we should avoid using "leaf" as it's always misleading (it's irrelevant from the perspective of the proven shape whether a node is a leaf or an inner node of the original structure). Then we could use "requested" and "helper", and "witness" as the union of the two.

Reply:
Ok my bad. I thought witness was synonymous with "the helper nodes".